src/index/index.c

```diff
@@ -0,0 +1,214 @@
+  Index *index = malloc(sizeof(Index));
+
+  index->label = strdup(label);
+  index->property = strdup(prop_str);
+
+  initializeSkiplists(index);
+
+  TuplesIter *it = TuplesIter_new(label_matrix);
+
+  Node *node;
+  EntityProperty *prop;
+
+  skiplist *sl;
+  GrB_Index node_id;
+  int found;
+  int prop_index = 0;
+  while(TuplesIter_next(it, NULL, &node_id) != TuplesIter_DEPLETED) {
+    node = Graph_GetNode(g, node_id);
+    // If the sought property is at a different offset than it occupied in the previous node,
+    // then seek and update
+    if (strcmp(prop_str, node->properties[prop_index].name)) {
+      found = 0;
+      for (int i = 0; i < node->prop_count; i ++) {
+        prop = node->properties + i;
+        if (!strcmp(prop_str, prop->name)) {
+          prop_index = i;
+          found = 1;
+          break;
+        }
+      }
+    } else {
+      found = 1;
+    }
+    // The targeted property does not exist on this node
+    if (!found) continue;
+
+    prop = node->properties + prop_index;
+    // This value will be cloned within the skiplistInsert routine if necessary
+    SIValue *key = &prop->value;
+
+    assert(key->type == T_STRING || key->type & SI_NUMERIC);
+    sl = (key->type & SI_NUMERIC) ? index->numeric_sl: index->string_sl;
+    skiplistInsert(sl, key, node_id);
+  }
+
+  TuplesIter_free(it);
+
+  return index;
+}
+
+// Create and populate index for specified property
+// (This function will create separate string and numeric indices if property has mixed types)
+void Index_Create(RedisModuleCtx *ctx, const char *graphName, Graph *g, const char *label, const char *prop_str) {
+  // Set up a reply of just statistics
+  RedisModule_ReplyWithArray(ctx, 1);
+  RedisModule_ReplyWithArray(ctx, 2);
+
+  RedisModuleKey *key = _index_LookupKey(ctx, graphName, label, prop_str);
+  // Do nothing if this key already exists
+  if (RedisModule_ModuleTypeGetType(key) != REDISMODULE_KEYTYPE_EMPTY) {
+    RedisModule_CloseKey(key);
+    RedisModule_ReplyWithSimpleString(ctx, "(no changes, no records)");
+    return;
+  }
+
+  LabelStore *store = LabelStore_Get(ctx, STORE_NODE, graphName, label);
+  assert(store);
+  const GrB_Matrix label_matrix = Graph_GetLabelMatrix(g, store->id);
+
+  Index *idx = buildIndex(g, label_matrix, label, prop_str);
+  RedisModule_ModuleTypeSetValue(key, IndexRedisModuleType, idx);
+  RedisModule_CloseKey(key);
+
+  RedisModule_ReplyWithSimpleString(ctx, "Added 1 index.");
+}
+
+/* Output text for EXPLAIN calls */
+char* Index_OpPrint(AST_IndexNode *indexNode) {
+  switch(indexNode->operation) {
+    case CREATE_INDEX:
+      return "Create Index";
+    default:
+      return "Drop Index";
+  }
+}
+
+/* Generate an iterator with no lower or upper bound. */
+IndexIter* IndexIter_Create(Index *idx, SIType type) {
+  skiplist *sl = type == T_STRING ? idx->string_sl : idx->numeric_sl;
+  return skiplistIterateAll(sl);
+}
+
+/* Apply a filter to an iterator, returns 1 if filter can be deleted */
```

Comment:
Please rephrase: "returns 1 if filter can be deleted"
We won't always be in the context/mindset of SCAN op reduction.